File: quantum-encryptor-gateway/src/main/java/com/quantum/gateway/util/HexUtil.java
package com.quantum.gateway.util;

import java.security.SecureRandom;

public final class HexUtil {

    private static final SecureRandom SECURE_RANDOM = new SecureRandom();

    private HexUtil() {}

    public static String bytesToHex(byte[] bytes) {
        StringBuilder sb = new StringBuilder();
        for (byte b : bytes) {
            sb.append(String.format("%02x", b));
        }
        return sb.toString();
    }

    public static byte[] hexToBytes(String hex) {
        if (hex == null || hex.length() % 2 != 0) {
            throw new IllegalArgumentException("Invalid hex string");
        }
        int len = hex.length();
        byte[] data = new byte[len / 2];
        for (int i = 0; i < len; i += 2) {
            data[i / 2] = (byte) ((Character.digit(hex.charAt(i), 16) << 4)
                    + Character.digit(hex.charAt(i + 1), 16));
        }
        return data;
    }

    public static String generateRandomHex(int byteLength) {
        byte[] bytes = new byte[byteLength];
        SECURE_RANDOM.nextBytes(bytes);
        return bytesToHex(bytes);
    }
}

File: quantum-client-android/app/src/main/java/com/quantum/poc/util/HexUtil.kt
package com.quantum.poc.util

import java.nio.charset.StandardCharsets

object HexUtil {
    
    fun bytesToHex(bytes: ByteArray): String {
        return bytes.joinToString("") { "%02x".format(it) }
    }
    
    fun hexToBytes(hex: String): ByteArray {
        return hex.chunked(2).map { it.toInt(16).toByte() }.toByteArray()
    }
    
    fun stringToHex(str: String): String {
        return bytesToHex(str.toByteArray(StandardCharsets.UTF_8))
    }
    
    fun hexToString(hex: String): String {
        return String(hexToBytes(hex), StandardCharsets.UTF_8)
    }
}
